File: problem230.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        heap = []
        
        def traverse(node):
            nonlocal heap, k
            if not node:
                return
            heapq.heappush(heap, -node.val)
            if len(heap)>k:
                heapq.heappop(heap)
                
            traverse(node.left)
            traverse(node.right)
            
        heapq.heappush(heap, -root.val)
        traverse(root.left)
        if len(heap)<k:
            traverse(root.right)
            
        return -heap[0]
        
#Time: O(Nlogk)
#Space: O(k)

File: problem230.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 
 
- traverse the entire tree
- use the property of BST
- 
 
 */

public class Solution {
    int k;
    int smallest;
    public int KthSmallest(TreeNode root, int k) {
        this.k=k;
        Traverse(root);
        return smallest;
    }
    
    public void Traverse(TreeNode node)
    {
        
        if(node==null || k<1)
            return;
        
        Traverse(node.left);
        
        if(k-- == 1)
        {
            smallest=node.val;
            return;
        }
        
        Traverse(node.right);
    }
}

//Time: O(K)
//Space: O(1)
